src/librustc/middle/def_id.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright 2012-2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use ast_map;
+use middle::ty;
+use syntax::ast::{CrateNum, NodeId};
+use std::fmt;
+
+#[derive(Clone, Eq, Ord, PartialOrd, PartialEq, RustcEncodable,
+           RustcDecodable, Hash, Copy)]
+pub struct DefId {
+    pub krate: CrateNum,
+    pub node: NodeId,
+}
+
+impl fmt::Debug for DefId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        try!(write!(f, "DefId {{ krate: {}, node: {}",
+                    self.krate, self.node));
+
+        // Unfortunately, there seems to be no way to attempt to print
+        // a path for a def-id, so I'll just make a best effort for now
+        // and otherwise fallback to just printing the crate/node pair
+        try!(ty::tls::with_opt(|opt_tcx| {
+            if let Some(tcx) = opt_tcx {
+                if self.is_local() {
+                    match tcx.map.find(self.node) {
+                        Some(ast_map::NodeItem(..)) |
+                        Some(ast_map::NodeForeignItem(..)) |
+                        Some(ast_map::NodeImplItem(..)) |
+                        Some(ast_map::NodeTraitItem(..)) |
+                        Some(ast_map::NodeVariant(..)) |
+                        Some(ast_map::NodeStructCtor(..)) => {
+                            return write!(f, " => {}", tcx.item_path_str(*self));
```

Comment:
@eddyb 

> I'm not sure there's even a point to restricting this to only certain nodes anymore, if there ever was.

perhaps not; looking over the code it seems like it will always gin up *some* kind of path, even if it's an empty or useless one.